### PR TITLE
Add ModelProto support for quantize api

### PR DESCRIPTION
**Pull request by xiaoyu-work**
### Description
Add ModelProto support for `quantize` api

### Motivation and Context
Currently, the `quantize` API only accepts a model path as the input model. However, for large models, saving and loading from disk can be time-consuming. By adding `ModelProto` as an input option to the `quantize` API, significant time can be saved.

### CODE REVIEW

File: onnxruntime/python/tools/quantization/quantize.py [hunk 1]
```diff
@@ -479,14 +482,14 @@ def inc_dataloader():
...
+        sq_model_path = Path(sq_path.name).joinpath("sq_model.onnx").as_posix()
+        model.save(sq_model_path)
         nodes_to_exclude.extend([i.name for i in model.model.graph.node if i.name not in orig_nodes])
-        model = load_model_with_shape_infer(Path(model_input))  # use smooth quant model for calibration
+        model = load_model_with_shape_infer(Path(sq_model_path))  # use smooth quant model for calibration
 
     with tempfile.TemporaryDirectory(prefix="ort.quant.") as quant_tmp_dir:
         calibrator = create_calibrator(
-            Path(model_input),
+            Path(sq_model_path),
```
Comment (yufenglee) [inline onnxruntime/python/tools/quantization/quantize.py:492]: if smoothQuant is false, there will be no a sq_model_path

Comment (xiaoyu-work) [reply onnxruntime/python/tools/quantization/quantize.py:492]: Yeah that's true I missed that. I'll change it back.

File: onnxruntime/python/tools/quantization/quantize.py [hunk 2]
```diff
@@ -6,6 +6,10 @@
 import logging
 import tempfile
 from pathlib import Path
+from typing import Union
+
+import onnx
+from onnx_model import ONNXModel
```
Comment (tianleiwu) [inline onnxruntime/python/tools/quantization/quantize.py:12]: from .onnx_model

File: onnxruntime/python/tools/quantization/quantize.py [hunk 3]
```diff
@@ -435,7 +439,7 @@ def quantize_static(
         qdq_ops = list(QDQRegistry.keys())
         op_types_to_quantize = list(set(q_linear_ops + qdq_ops))
 
-    model = load_model_with_shape_infer(Path(model_input))
+    model = model_input if isinstance(model_input, onnx.ModelProto) else load_model_with_shape_infer(Path(model_input))
```
Comment (tianleiwu) [inline onnxruntime/python/tools/quantization/quantize.py:442]: do shape inference when model_input is ModelProto?

File: onnxruntime/python/tools/quantization/quantize.py [hunk 4]
```diff
@@ -485,6 +489,12 @@ def inc_dataloader():
         model = load_model_with_shape_infer(Path(model_input))  # use smooth quant model for calibration
 
     with tempfile.TemporaryDirectory(prefix="ort.quant.") as quant_tmp_dir:
+        if isinstance(model_input, onnx.ModelProto):
```
Comment (tianleiwu) [inline onnxruntime/python/tools/quantization/quantize.py:492]: How about Line 482: does ORTSmoothQuant function could handle ModelProto input?

File: onnxruntime/python/tools/quantization/quantize.py [hunk 5]
```diff
@@ -609,7 +619,7 @@ def quantize_dynamic(
     if not op_types_to_quantize or len(op_types_to_quantize) == 0:
         op_types_to_quantize = list(IntegerOpsRegistry.keys())
 
-    model = load_model_with_shape_infer(Path(model_input))
+    model = model_input if isinstance(model_input, onnx.ModelProto) else load_model_with_shape_infer(Path(model_input))
```
Comment (tianleiwu) [inline onnxruntime/python/tools/quantization/quantize.py:622]: Same here: skip shape inference might have negative impact on quantization?

File: onnxruntime/python/tools/quantization/quantize.py (hunk 4 above)
Comment (xiaoyu-work) [reply onnxruntime/python/tools/quantization/quantize.py:492]: Yes I've checked that `ORTSmoothQuant` supports ModelProto

File: onnxruntime/python/tools/quantization/quantize.py (hunk 3 above)
Comment (xiaoyu-work) [reply onnxruntime/python/tools/quantization/quantize.py:442]: Good call. I'll investigate more on this.

File: onnxruntime/python/tools/quantization/quantize.py (hunk 5 above)
Comment (xiaoyu-work) [reply onnxruntime/python/tools/quantization/quantize.py:622]: Same as above.

File: onnxruntime/python/tools/quantization/quantize.py (hunk 3 above)
Comment (xiaoyu-work) [reply onnxruntime/python/tools/quantization/quantize.py]: Updatedd

File: onnxruntime/python/tools/quantization/execution_providers/qnn/preprocess.py [hunk 6]
```diff
@@ -16,8 +17,8 @@
 
 
 def qnn_preprocess_model(
-    model_input: Path,
-    model_output: Path,
+    model_input: Union[str, Path, onnx.ModelProto],
```
[the remark this answers is not in the document]
Comment (xiaoyu-work) [reply onnxruntime/python/tools/quantization/execution_providers/qnn/preprocess.py:20]: Any reason why ruff only checks this file for type annotations?

File: onnxruntime/python/tools/quantization/quantize.py (hunk 4 above)
Comment (xiaoyu-work) [reply onnxruntime/python/tools/quantization/quantize.py:496]: - https://github.com/intel/neural-compressor/blob/d80ab264d6d3995f95d9c562a6ef91ce239b322a/neural_compressor/adaptor/ox_utils/smooth_quant.py#L120
- https://github.com/intel/neural-compressor/blob/d80ab264d6d3995f95d9c562a6ef91ce239b322a/neural_compressor/model/onnx_model.py#L42

File: onnxruntime/python/tools/quantization/shape_inference.py [hunk 7]
```diff
@@ -9,11 +9,12 @@
 import tempfile
 import traceback
 from pathlib import Path
-from typing import Optional
+from typing import Optional, Union
 
 import onnx
 
 import onnxruntime
+from onnxruntime.python.tools.transformers.onnx_utils import extract_raw_data_from_model, has_external_data
```
Comment (tianleiwu) [inline onnxruntime/python/tools/quantization/shape_inference.py:17]: from onnxruntime.transformers.onnx_utils

File: onnxruntime/python/tools/quantization/shape_inference.py [hunk 8]
```diff
@@ -109,34 +123,45 @@ def quant_pre_process(
...
+            input_model = opt_model_path
 
         if not skip_onnx_shape:
             # ONNX shape inference.
             # According to docs, infer_shapes_path should be used for 2G+ models.
             # If the skip optimization is specified, we could be dealing with a
             # large model. So be on the safe side, save the model
             if model is not None:
-                input_model_path = str(temp_path / "symbolic_shape_inferred.onnx")
+                input_model = str(temp_path / "symbolic_shape_inferred.onnx")
```
Comment (tianleiwu) [inline onnxruntime/python/tools/quantization/shape_inference.py:134]: nit: This is not symbolic shape info, may use a different name like "onnx_shape_inferred.onnx", but this does not impact result.

**Review by tianleiwu — APPROVED** [review]

**Pull request merged**